finality-aleph/src/data_io/data_provider.rs

```diff
@@ -0,0 +1,266 @@
+            if curr_block.num - finalized_block.num
+                <= <NumberFor<B>>::saturated_from(MAX_DATA_BRANCH_LEN)
+            {
+                branch.push(curr_block.hash);
+            }
+            curr_block = get_parent(self.client.clone(), &curr_block)
+                .expect("block of num >= 1 must have a parent")
+        }
+        if curr_block.hash == finalized_block.hash {
+            let num_last = finalized_block.num + <NumberFor<B>>::saturated_from(branch.len());
+            // The hashes in `branch` are ordered from top to bottom -- need to reverse.
+            branch.reverse();
+            *self.data_to_propose.lock().await =
+                AlephData::HeadProposal(AlephUnvalidatedProposal::<B>::new(branch, num_last));
+        } else {
+            // By backtracking from the best block we reached a block conflicting with best finalized.
+            // This is most likely a bug, or some extremely unlikely synchronization issue of the client.
+            warn!(target: "afa", "Error updating data. Conflicting blocks: {:?}, finalized {:?}", curr_block, finalized_block);
+            // We leave the old data intact -- it is not up-to-date, but at least it is correct.
+        }
+    }
+
+    pub(crate) async fn get_best_header(&self) -> B::Header {
+        self.select_chain.best_chain().await.expect("No best chain")
+    }
+
+    pub(crate) async fn run(mut self, mut exit: oneshot::Receiver<()>) {
+        // We would like `best_block_in_session` to correspond to the highest ancestor of the `best_block` (this is what
+        // `select_chain` provides us with) up to the maximal height of `max_block_num`. This task periodically
+        // queries `select_chain` for the `best_block` and updates `best_block_in_session` accordingly. One optimization that it
+        // uses is that once the block in `best_block_in_session` reaches the height of `max_block_num`, and the just queried
+        // `best_block` is a `descendant` of the previous query, then we don't need to update `best_block_in_session`, as it is
+        // already correct. After updating `best_block_in_session`, we call `update_data` to update `data_to_propose`.
+        let max_block_num = self.session_boundaries.last_block();
+        let mut prev_best_block: Option<BlockHashNum<B>> = None;
+        loop {
+            let delay = futures_timer::Delay::new(Duration::from_millis(REFRESH_INTERVAL));
+            tokio::select! {
+                _ = delay => {
+                    let new_best_header = self.get_best_header().await;
+                    let new_best_block: BlockHashNum<B> = (new_best_header.hash(), *new_best_header.number()).into();
+                    if new_best_header.number() <= &max_block_num {
+                        self.best_block_in_session = Some(new_best_block.clone());
+                    } else if prev_best_block.is_none() || prev_best_block.as_ref().unwrap().num < max_block_num {
```

Comment:
I don't know about whitespace, this is what `fmt` outputs ;/

Regarding the unwraps: I think the current version is better, as it clearly separates the cases that should be handled distinctly and just does that. You version would introduce duplication.

Reply:
If there are whitespace issues then it's on `fmt` :P 

Otherwise, I have changed it to `match` -- which I don't think made it more readable, but that's never objective. Anyway, done.

Reply:
Not sure about the `match` either, but the comments got somewhat better and this is now very readable. <3